awx/main/notifications/webhook_backend.py

```diff
@@ -61,23 +63,60 @@ def format_body(self, body):
 
     def send_messages(self, messages):
         sent_messages = 0
+        retries = 0
         if self.http_method.lower() not in ['put', 'post']:
             raise ValueError("HTTP method must be either 'POST' or 'PUT'.")
         chosen_method = getattr(requests, self.http_method.lower(), None)
+        resp = None
+        '''
+        Sometimes we hit redirect URLs. We must account for this. We still extract the redirect URL from the response headers and try again. Max retires == 5
+        '''
         for m in messages:
+
             auth = None
             if self.username or self.password:
                 auth = (self.username, self.password)
-            r = chosen_method(
-                "{}".format(m.recipients()[0]),
-                auth=auth,
-                data=json.dumps(m.body, ensure_ascii=False).encode('utf-8'),
-                headers=dict(list(get_awx_http_client_headers().items()) + list((self.headers or {}).items())),
-                verify=(not self.disable_ssl_verification),
-            )
-            if r.status_code >= 400:
-                logger.error(smart_str(_("Error sending notification webhook: {}").format(r.status_code)))
-                if not self.fail_silently:
-                    raise Exception(smart_str(_("Error sending notification webhook: {}").format(r.status_code)))
-            sent_messages += 1
+
+            # convert emtpy string to empty dict for POST/PUT
+            if not m.body:
+                m.body = dict()
+
+            url = "{}".format(m.recipients()[0])
+            data = json.dumps(m.body, ensure_ascii=False).encode('utf-8')
+
+            headers = dict(list(get_awx_http_client_headers().items()) + list((self.headers or {}).items()))
+
+            for retries in range(self.MAX_RETIRES):
+
+                resp = chosen_method(
+                    url=url,
+                    auth=auth,
+                    data=data,
+                    headers=headers,
+                    verify=(not self.disable_ssl_verification),
+                    allow_redirects=False,  # override default behaviour for redirects
+                )
+
+                # we've hit a redirect. extract the redirect URL out of the first response header and try again
+                if resp.status_code in [301, 307]:
+                    # take the first redirect URL in the response header and try that
+                    url = resp.headers.get("Location", None)
+                    if url is None:
+                        raise Exception(smart_str(_("Recieved redirect to a blank URL. Response headers={}").format(resp_inner.history[0].headers)))
+                    continue
+
+                if resp.status_code in range(200, 299):
```

Comment:
I think the loop logic could be a bit more compact:

1. This block goes away entirely.
2. The redirect status code conditional above gets an `else:` clause with just a `break` in it.
3. The error status code conditional below goes outside of the loop.
4. A new 'skinny' redirect code conditional also goes outside of the loop, to catch the case where we exceeded the retries.

Right?  Both a successful request and an unsuccessful request are terminating conditions, so the loop is no longer necessary at that point.  And putting the error code check and a final redirect code check outside of the loop removes the need to track the message count.

Alternatively, you could use the admittedly difficult to understand `else:` clause of `for` loops to avoid the outer check for a redirect code:

```python
for attempt in range(self.RETRIES):
    # ...
else:
    # The `break` condition was never hit inside the loop, so we exceeded the retries.
    logger.error(...)
    raise Exception(...) 
```